scripts/!adetailer.py

```diff
@@ -668,6 +668,79 @@ def get_image_mask(p) -> Image.Image:
             width, height = p.width, p.height
         return images.resize_image(p.resize_mode, mask, width, height)
 
+    @staticmethod
+    def get_dynamic_denoise_strength(denoise_strength, bbox, image):
+        denoise_power = opts.data.get("ad_dynamic_denoise_power", 0)
+        if denoise_power == 0:
+            return denoise_strength
+
+        image_pixels = image.width * image.height
+        bbox_pixels = (bbox[2] - bbox[0]) * (bbox[3] - bbox[1])
+
+        normalized_area = bbox_pixels / image_pixels
+        denoise_modifier = (1.0 - normalized_area) ** denoise_power
+
+        print(
+            f"[-] ADetailer: dynamic denoising -- {denoise_modifier:.2f} * {denoise_strength:.2f} = {denoise_strength * denoise_modifier:.2f}"
+        )
+
+        return denoise_strength * denoise_modifier
+
+    @staticmethod
+    def get_optimal_crop_image_size(inpaint_width, inpaint_height, bbox):
+        calculate_optimal_crop = opts.data.get("ad_match_inpaint_bbox_size", False)
+        if not calculate_optimal_crop:
+            return (inpaint_width, inpaint_height)
+
+        resolutions = []
+        if shared.sd_model.is_sdxl:
+            # Limit resolutions to those SDXL was trained on.
+            resolutions = [
+                (1024, 1024),
+                (1152, 896),
+                (896, 1152),
+                (1216, 832),
+                (832, 1216),
+                (1344, 768),
+                (768, 1344),
+                (1536, 640),
+                (640, 1536),
+            ]
+        else:
+            msg = "[-] ADetailer: inpaint bounding box size matching is only valid for SDXL."
+            print(msg)
+            return (inpaint_width, inpaint_height)
+
+        bbox_width = bbox[2] - bbox[0]
+        bbox_height = bbox[3] - bbox[1]
+
+        target_aspect = bbox_width / bbox_height
+        larger_resolutions = [
+            res for res in resolutions if res[0] >= bbox_width and res[1] >= bbox_height
+        ]
+
+        if not larger_resolutions:
+            return (inpaint_width, inpaint_height)
+
+        def aspect_ratio_difference(res):
+            res_aspect = res[0] / res[1]
+            return abs(res_aspect - target_aspect)
+
+        optimal_resolution = min(larger_resolutions, key=aspect_ratio_difference)
+
+        # Don't use resolution if it's smaller than what was to be used.
+        if (
+            optimal_resolution[0] < inpaint_width
+            and optimal_resolution[1] < inpaint_height
+        ):
+            return (inpaint_width, inpaint_height)
```

Comment:
How about reversing the order by keeping only candidates that are equal to or larger than the inpaint height and width first, and then finding the optimal resolution?

Reply:
This is more of a sanity check than anything else, to make sure we don't somehow return an inpaint area smaller than what was originally going to be used. We also need all the SDXL resolutions up front, as we're trying to find one that best fits the aspect ratio of the bounding box. SDXL resolutions shuffle around 64 pixels between the width and height, so if we filter on `>=` alone, we'll end up with no valid resolutions (or just the inpaint resolution itself) to match against most of the time.

That said, I can certainly adjust the early filter to make sure the resolution candidates are at least bigger in one dimension, which should make this check unnecessary.

Reply:
I'm also toying with the idea of a "strict" mode (current implementation) and a "free" mode that just matches the aspect ratio and would work with any model. Inference seems a lot more forgiving of odd resolutions late in the denoising process / low denoise strengths, so it *might* be OK it most cases.

Reply:
OK, I've added the alternate modes. As per the commit comments, strict is the same as the original implementation and for SDXL only, while the two free modes should work with anything. I added a "prefer smaller" and "prefer larger" options, as my initial hunch of using the larger dimension usually resulted in larger inpaint sizes that deviated too much from the original inpaint sizes. Smaller is more conservative and almost always is a closer match for the original dimensions.

If you don't have any other suggested changes, this should be good to merge!